keps/sig-node/3066-subsecond-probes/README.md

```diff
@@ -0,0 +1,644 @@
+#### Knative
+
+For example, Knative will create Pods (via Deployment) and wait for them to become `Ready`
+when handling an HTTP request from an end-user. In this case, Pod readiness
+latency has a direct impact on HTTP request latency. (In the steady state,
+Knative will re-use an existing Pod, but this situation can happen on scale-up,
+and is guaranteed to happen on scale-from-zero.)
+
+### Goals
+
+An ability to specify probe durations that have more granularity, more specificity, than just zero,
+one, or n-seconds. Focus is on possibly faster readiness probes, for example every .5 seconds until succss
+and more granular use beyond existing defaults, for example 1.5 second duration instead of picking
+between 1 or 2 second duration.
+Add additional tests cases to the timeout test cases.
+Not breaking backwards compatibility with the V1 API.
+
+
+### Non-Goals
+
+V2 API for existing objects.
+Converting fields from `int32` to `resource.Quantity`.
+
+
+## Proposal
+
+TLDR:
+Add two new optional fields (of type `*int32`) to the Probe struct, which would be used to offset the second based time values in the Probe struct:
+
+- `PeriodOffsetMilliseconds`       // How often (in milliseconds) to offset PeriodSeconds when performing the probe (*** to reduce un-necessary resource usage, kubelet may provide means for mitigation of resource impacts; for example, the offset may only be applied until success then no longer. Another example, kubelet may increase the period duration). Values must be between [-999,999], and the effective sum of `PeriodSeconds` and `PeriodOffsetMilliseconds` must be greater than 0.
+- `InitialOffsetDelayMilliseconds` // Length of time (in milliseconds) to offset IntialDelaySeconds before health checking is activated. Values must be between [-999,999], and the effective sum of `InitialDelaySeconds` and `InitialOffsetDelayMilliseconds` must be greater than or equal to 0.
+
+The seconds and milliseconds fields (as related) will be summed to get the resulting time duration. For example,
+
+```yaml
+...
+#yaml
+periodSeconds: 1
+periodOffsetMilliseconds: 500
+...
+```
+
+would be considered a period value of 1.5s or 1500ms. These values are used as `time.Duration` within the prober package ([ex1](https://github.com/kubernetes/kubernetes/blob/a750d8054a6cb3167f495829ce3e77ab0ccca48e/pkg/kubelet/prober/prober.go#L161) [ex2](https://github.com/kubernetes/kubernetes/blob/a750d8054a6cb3167f495829ce3e77ab0ccca48e/pkg/kubelet/prober/worker.go#L132) ), so there's no real difference between 1.5s and 1500ms.
+
+A few additional examples:
+
+`periodSeconds: 2` and `periodOffsetMilliseconds: -500` would be 1.5s / 1500ms.
+
+`periodSeconds: 1` and `periodOffsetMilliseconds: -500` would be 0.5s / 500ms. (*** To reduce un-necessary resource usage, when periodOffsetMilliseconds is used to reduce the period to less than a second, the resulting duration will not be guaranteed by kubelet. Kubelet may adjust the offset based on available resources.)
```

Comment:
Not blocking - Not sure how much we can quantify available resources. I suspect at some point we may end up needing a probe quota for a node (especially for heavier exec probes.)

Reply:
nod... incorporating